packages/flutter/lib/src/material/ink_sparkle.dart

```diff
@@ -0,0 +1,765 @@
+  return math.max(d1, d2) / 2.0;
+}
+
+  /// Begin an m3 sparkle ripple, centered at [position] relative to
+  /// [referenceBox].
+  ///
+  /// The [controller] argument is typically obtained via
+  /// `Material.of(context)`.
+  ///
+  /// If [containedInkWell] is true, then the ripple will be sized to fit
+  /// the well rectangle, then clipped to it when drawn. The well
+  /// rectangle is the box returned by [rectCallback], if provided, or
+  /// otherwise is the bounds of the [referenceBox].
+  ///
+  /// If [containedInkWell] is false, then [rectCallback] should be null.
+  /// The ink ripple is clipped only to the edges of the [Material].
+  /// This is the default.
+  ///
+  /// When the ripple is removed, [onRemoved] will be called.
+class InkSparkle extends InteractiveInkFeature {
+  InkSparkle({
+    required MaterialInkController controller,
+    required RenderBox referenceBox,
+    required Offset position,
+    required Color color,
+    required TextDirection textDirection,
+    bool containedInkWell = true,
+    RectCallback? rectCallback,
+    BorderRadius? borderRadius,
+    ShapeBorder? customBorder,
+    double? radius,
+    VoidCallback? onRemoved,
+  })  : _color = color,
+        _rectCallback = rectCallback,
+        _position = position,
+        _borderRadius = borderRadius ?? BorderRadius.zero,
+        _customBorder = customBorder,
+        _textDirection = textDirection,
+        _targetRadius = radius ??
+            _getTargetRadius(
+                referenceBox, containedInkWell, rectCallback, position),
+        _clipCallback =
+            _getClipCallback(referenceBox, containedInkWell, rectCallback),
+        super(
+            controller: controller,
+            referenceBox: referenceBox,
+            color: color,
+            onRemoved: onRemoved) {
+    controller.addInkFeature(this);
+
+    FragmentShaderManager.inkSparkle().then((value) {
+      _shaderManager = value;
+    });
+
+    _ticker = controller.vsync.createTicker((elapsed) {
+      _time = elapsed;
+      controller.markNeedsPaint();
+    });
+    _ticker.start();
+  }
+
+  late Ticker _ticker;
+  Duration _time = Duration.zero;
+  int _timeUpMillis = 0;
+
+  FragmentShaderManager? _shaderManager;
+  final Color _color;
+  final RectCallback? _rectCallback;
+
+  final Offset _position;
+  final BorderRadius _borderRadius;
+  final ShapeBorder? _customBorder;
+  final double _targetRadius;
+  final RectCallback? _clipCallback;
+  final TextDirection _textDirection;
+
+  /// Used to specify this type of ink splash for an [InkWell], [InkResponse],
+  /// material [Theme], or [ButtonStyle].
+  static const InteractiveInkFeatureFactory splashFactory = InkSparkleFactory();
+
+  @override
+  void confirm() {
+    // This method is only called on press up.
+    _timeUpMillis = _time.inMilliseconds;
+    // TODO(clocksmith): Verify value for ripple settle time.
+    Timer(const Duration(milliseconds: 150), () {
```

Comment:
I think it's convention to store constants at the top

Reply:
moved to top of class as static consts per flutter style guide